driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java

```diff
@@ -0,0 +1,188 @@
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(null, null);
+        });
+    }
+
+    /**
+     * See {@link #thenRunAndFinish(Runnable, SingleResultCallback)}, but the runnable
+     * will always be executed, including on the exceptional path.
+     * @param runnable the runnable
+     * @param callback the callback
+     */
+    default void thenAlwaysRunAndFinish(final Runnable runnable, final SingleResultCallback<Void> callback) {
+        this.finish((r, e) -> {
+            try {
+                runnable.run();
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(r, e);
+        });
+    }
+
+    /**
+     * @param runnable The async runnable to run after this one
+     * @return the composition of this and the runnable
+     */
+    default AsyncRunnable thenRun(final AsyncRunnable runnable) {
+        return (c) -> {
+            this.finish((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    runnable.finish(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param supplier The supplier to supply using after this runnable.
+     * @return the composition of this runnable and the supplier
+     * @param <T> The return type of the supplier
+     */
+    default <T> AsyncSupplier<T> thenSupply(final AsyncSupplier<T> supplier) {
+        return (c) -> {
+            this.finish((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    supplier.finish(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param errorCheck A check, comparable to a catch-if/otherwise-rethrow
+     * @param runnable   The branch to execute if the error matches
+     * @return The composition of this, and the conditional branch
+     */
+    default AsyncRunnable onErrorRunIf(
+            final Predicate<Throwable> errorCheck,
+            final AsyncRunnable runnable) {
+        return (callback) -> this.finish((r, e) -> {
+            if (e == null) {
+                callback.onResult(r, null);
+                return;
+            }
+            try {
+                boolean check = errorCheck.test(e);
+                if (check) {
+                    runnable.finish(callback);
+                    return;
+                }
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(r, e);
+        });
```

Comment:
To avoid writing the implementation of `onError...If` method twice, I suggest reusing the one on `AsyncSupplier`

```suggestion
        asSupplier().onErrorSupplyIf(errorCheck, runnable.asSupplier());
```